Implement D2k thumper

reaperrr opened the pull request:
Thumper replaces Medic.

`AttractsWorms` now only attracts them, while actors with `WormFodder` also get eaten.

reaperrr committed:
Implement D2k thumper

abcdefg30 commented on mods/d2k/rules/infantry.yaml:
```diff
@@ -70,32 +70,46 @@ trooper:
-		Name: Medic
-		Description: Heals nearby infantry\n  Strong vs Nothing\n  Weak vs Everything
+		Name: Thumper
+		Description: Attracts nearby worms\n  Strong vs Nothing\n  Weak vs Everything
```
`Description: Attracts nearby worms\n  Unarmed`

abcdefg30 commented on mods/d2k/rules/infantry.yaml:
```diff
@@ -70,32 +70,46 @@ trooper:
+		Facing: 128
+		AllowedTerrainTypes: Sand, Spice, Dune
+		DeploySound:
+		UndeploySound:
```
Both sounds already default to `null`?

pchote commented on mods/d2k/rules/infantry.yaml:
```diff
@@ -70,32 +70,45 @@ trooper:
-		AttackSequence: heal
+		Speed: 43
+	DeployToUpgrade:
+		Upgrades: deployed
```
We'll also want an `AmbientSound` definition for the thump audio.

reaperrr replied:
Done.

reaperrr committed:
Make AmbientSound upgradable and add customisable interval

reaperrr committed:
Add D2k thumper sound

reaperrr committed:
Align thumper-sand overlay with thump anim

GraionDilach commented:
I have a feeling that ambientsounds with Interval=0 cannot be controlled via disabling the trait.

Phrohdoh commented on OpenRA.Mods.D2k/Traits/AttractsWorms.cs:
```diff
@@ -20,10 +20,10 @@ public class AttractsWormsInfo : UpgradableTraitInfo
 
 		[Desc("Noise percentage at Range step away from the actor.")]
-		public readonly int[] Falloff = { 100, 100, 25, 11, 6, 4, 3, 2, 1, 0 };
+		public readonly int[] Falloff = { 100, 100, 25, 12, 6, 3, 2, 1, 0 };
```
At this point is it safe to change default values (which mods may rely on), instead of changing our yaml?

reaperrr commented:
Updated.

MicroJOo commented:
works as expected :) ... :+1:

problem though is that you can trap him "forever" ... basicly take him out of a game
a solution could be , that he can get used only for a limited time ... and needs to recharge then
thats just an idea , of course that would be not true to the original ... but i expect the current implementation of the thumper to cause controversy in the future ... sooner or later 

however , this is really close to the original , i like it

Mailaender commented on OpenRA.Mods.D2k/Traits/Sandworm.cs:
```diff
@@ -99,7 +99,7 @@ void RescanForTargets(Actor self)
 
 			// If close enough, we don't care about other actors.
-			var target = self.World.FindActorsInCircle(self.CenterPosition, Info.IgnoreNoiseAttackRange).FirstOrDefault(x => x.Info.HasTraitInfo<AttractsWormsInfo>());
+			var target = self.World.FindActorsInCircle(self.CenterPosition, Info.IgnoreNoiseAttackRange).FirstOrDefault(x => x.Info.HasTraitInfo<WormFodderInfo>());
```
@pchote suggested to use target types instead. This would become `.FirstOrDefault(x => attackTrait.HasAnyValidWeapons(Target.FromActor(x)));` then.

Mailaender commented on mods/d2k/sequences/infantry.yaml:
```diff
@@ -196,10 +196,15 @@ medic: # actually thumper
+	thump: DATA.R8
 		Start: 1458
 		Length: 5
 		Tick: 480
```
Something isn't right here. I think you are missing `Transpose: true` because on your branch the deployed thumper is still holding his (other?) thumping device:

![image](https://cloud.githubusercontent.com/assets/756669/10749607/99713338-7c6c-11e5-8500-749b2774b862.png)

However I believe you want it to look like

![image](https://cloud.githubusercontent.com/assets/756669/10749584/5a990c26-7c6c-11e5-8353-1adec3df2df2.png)

compare https://github.com/OpenRA/OpenRA/pull/9019/.

penev92 replied:
This screenshot suggests that probably the idle animation isn't removed when deploying.

reaperrr replied:
Ah, my bad. The `@DEFAULT` added to `WithInfantryBody` meant it didn't override the `WithInfantryBody` inherited from the infantry default.

reaperrr committed:
Make sandworms not eat infantry

reaperrr commented:
Rebased and updated.

obrakmann commented:
Looks and works fine. Just two comments:

1. I think there's a typo in the second to last commit's commit message: "Align thumper _sand_..."

2. To address @GraionDilach's comment regarding controlling looped AmbientSounds via upgrades, I suggest the following code. I think that should fix it:

```csharp
class AmbientSound : UpgradableTrait<AmbientSoundInfo>, ITick
{
	ISound currentSound;
	bool wasDisabled = true;
	int interval;

	public AmbientSound(Actor self, AmbientSoundInfo info)
		: base(info)
	{
		interval = info.Interval;
	}

	public void Tick(Actor self)
	{
		if (IsTraitDisabled)
		{
			Game.Sound.StopSound(currentSound);
			currentSound = null;
			wasDisabled = true;
			return;
		}

		if (wasDisabled && Info.Interval == 0)
		{
			if (self.OccupiesSpace != null)
				currentSound = Game.Sound.PlayLooped(Info.SoundFile, self.CenterPosition);
			else
				currentSound = Game.Sound.PlayLooped(Info.SoundFile);
		}

		wasDisabled = false;

		if (Info.Interval == 0)
			return;

		if (interval-- > 0)
			return;

		interval = Info.Interval;

		if (self.OccupiesSpace != null)
			Game.Sound.Play(Info.SoundFile, self.CenterPosition);
		else
			Game.Sound.Play(Info.SoundFile);
	}
}
```

reaperrr commented:
@obrakmann 
> I think there's a typo in the second to last commit's commit message: "Align thumper *sand*..."

Unless you mean the missing `-` I'm not sure what you mean, there is a `thumper-sand` overlay which I tried to better align with the `thump` animation by reordering the frames. But I agree the commit message is a bit unprecise, I'll reword it a little.

obrakmann commented:
Oh nevermind. I probably should have looked at the actual commit. I assumed it was supposed to say "sound" instead of "sand". It's fine then.

reaperrr commented:
Updated.

Only minor change compared to @obrakmann's code: I changed `Info.Interval == 0` to `Info.Interval <= 0` so `-1` or some bogus higher negative value will have the same effect.

obrakmann commented:
:+1:

penev92 commented:
:+1:

penev92 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/6a6de7f1f8c81a030cdb871e189caa984c8b7bf5...777757b055ea8c3f9dfb74a0d606ac1653fc82ea)
